tests/react/vanilla-utils/selectAtom.test.tsx

```diff
@@ -231,3 +231,35 @@ it('equality function works even if suspend', async () => {
   await findByText('bigValue: {"a":1,"b":2}')
   await findByText('littleValue: {"a":1}')
 })
+
+it.only('should not return async value when the base atom values are synchronous', async () => {
+  expect.assertions(4)
+  type Base = { id: number; value: number }
+  const initialBase = Promise.resolve({ id: 0, value: 0 })
+  const baseAtom = atom<Base | Promise<Base>>(initialBase)
+  const idAtom = selectAtom(
+    baseAtom,
+    ({ id }) => id,
```

Comment:
Or, I wonder if resolving promises inside selectAtom was a good idea...

Reply:
I was wondering that too. It would be a breaking change if we revert this behavior.

IMO, not resolving promises is not a very good DX. You still have the problem of promises being contagious

```ts
const baseAtom = atom(Promise.resolve({ id: 0, value: 0 }))

const derivedAsyncAtom = selectAtom(baseAtom,
  async (valueMaybePromise, sliceMaybePromise) => {
    const value = await valueMaybePromise
    const slice = await sliceMaybePromise
    ...do something with slice and value
  }
)
```

Reply:
> not a very good DX

That's what I thought too, but for `jotai/utils` we want straightforward impls (and open space for better DX for `jotaijs/*`.)

> It would be a breaking change if we revert this behavior.

Yeah, that's the problem. Not sure if we can provide a migration path (=only deprecate the behavior with warning, and remove it in the next major.)

Reply:
In this case, we should either:
A. create a weak map solution for replacing promises with resolved values
B. deprecate resolving promises inside selectAtom

What should we do?

Reply:
Yes, and at this point, both options sound valid. I would like to take B, depending on the hackiness with the implementation of A, which is likely?

Yeah, so my current proposition is B, which is the philosophy of Jotai core (along with atomWithDefault change #1939). Well, I think we can and did add breaking changes with minor bump for `jotai/utils`. https://github.com/pmndrs/jotai/releases/tag/v2.2.0
We assume a theoretical bug of `selectAtom` resolving promises.

Reply:
The weakmap solution is ready for your review. Please let me know your final decision here. I'm ok either way.

Reply:
for my own knowledge, what is the issue with a refAtom?
atomEffect uses a ref, if its an issue, I'll update it.

Reply:
It's not an issue, if it's done properly, but it's a hack and we want to avoid hacks as much as possible.